ledger/participant-integration-api/src/test/lib/scala/platform/indexer/ha/EndlessReadService.scala

```diff
@@ -0,0 +1,222 @@
+  */
+case class EndlessReadService(
+    updatesPerSecond: Int,
+    name: String,
+)(implicit loggingContext: LoggingContext)
+    extends ReadService
+    with AutoCloseable {
+  import EndlessReadService._
+
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  override def currentHealth(): HealthStatus =
+    if (aborted) HealthStatus.unhealthy else HealthStatus.healthy
+
+  override def ledgerInitialConditions(): Source[LedgerInitialConditions, NotUsed] = {
+    logger.info("EndlessReadService.ledgerInitialConditions() called")
+    initialConditionCalls.incrementAndGet()
+    Source
+      .single(LedgerInitialConditions(ledgerId, configuration, recordTime(0)))
+      .via(killSwitch.flow)
+  }
+
+  /** Produces the following stream of updates:
+    *    1. a config change
+    *    1. a party allocation
+    *    1. a package upload
+    *    1. a transaction that creates a contract
+    *    1. a transaction that archives the previous contract
+    *
+    *    The last two items above repeat indefinitely
+    */
+  override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] = {
+    logger.info(s"EndlessReadService.stateUpdates($beginAfter) called")
+    stateUpdatesCalls.incrementAndGet()
+    val startIndex: Int = beginAfter.map(index).getOrElse(1)
+    Source
+      .fromIterator(() => Iterator.from(startIndex))
+      .throttle(updatesPerSecond, 1.second)
+      .map {
+        case i @ 1 =>
+          offset(i) -> Update.ConfigurationChanged(
+            recordTime(i),
+            submissionId(i),
+            participantId,
+            configuration,
+          )
+        case i @ 2 =>
+          offset(i) -> Update.PartyAddedToParticipant(
+            party,
+            "Operator",
+            participantId,
+            recordTime(i),
+            Some(submissionId(i)),
+          )
+        case i @ 3 =>
+          offset(i) -> Update.PublicPackageUpload(
+            List(archive),
+            Some("Package"),
+            recordTime(i),
+            Some(submissionId(i)),
+          )
+        case i if i % 2 == 0 =>
+          offset(i) -> Update.TransactionAccepted(
+            optCompletionInfo = Some(completionInfo(i)),
+            transactionMeta = transactionMeta(i),
+            transaction = createTransaction(i),
+            transactionId = transactionId(i),
+            recordTime = recordTime(i),
+            divulgedContracts = List.empty,
+            blindingInfo = None,
+          )
+        case i =>
+          offset(i) -> Update.TransactionAccepted(
+            optCompletionInfo = Some(completionInfo(i)),
+            transactionMeta = transactionMeta(i),
+            transaction = exerciseTransaction(i),
+            transactionId = transactionId(i),
+            recordTime = recordTime(i),
+            divulgedContracts = List.empty,
+            blindingInfo = None,
+          )
+      }
+      .via(killSwitch.flow)
```

Comment:
shouldn't this method need to be synchronized? it accesses a variable

Reply:
Good point. Added `synchronized` to all methods.